bundles/umljava/tools.vitruv.applications.umljava.java2uml/src/tools/vitruv/applications/umljava/java2uml/JavaToUmlMethod.reactions

```diff
@@ -25,58 +26,63 @@ import routines javaToUmlTypePropagation using qualified names
 reaction JavaClassMethodCreated {
     after element java::ClassMethod created and inserted in java::ConcreteClassifier[members]
     with (affectedEObject instanceof Class)
-    	|| (affectedEObject instanceof Enumeration)
-    call createUmlClassMethod(newValue, affectedEObject)
+        || (affectedEObject instanceof Enumeration)
+    call createOrFindUmlClassMethod(newValue, affectedEObject)
 }
 
-routine createUmlClassMethod(java::ClassMethod jMeth, java::ConcreteClassifier jClassifier) {
-    match {
+routine createOrFindUmlClassMethod(java::ClassMethod jMethod, java::ConcreteClassifier jClassifier) {
+     match {
         val uClassifier = retrieve uml::Classifier corresponding to jClassifier
-		require absence of uml::Operation corresponding to jMeth
+            with uClassifier instanceof OperationOwner
+        require absence of uml::Operation corresponding to jMethod
     }
     action {
-        val uOperation = create uml::Operation and initialize {
-            uOperation.name = jMeth.name;
-        }
-        add correspondence between uOperation and jMeth
         call {
-            if (uClassifier instanceof org.eclipse.uml2.uml.Class) {
-                addUmlOperationToClass(uClassifier, uOperation)
-            } else if (uClassifier instanceof org.eclipse.uml2.uml.Enumeration) {
-                addUmlOperationToEnum(uClassifier, uOperation)
+            val foundMethod = (uClassifier as OperationOwner).ownedOperations.findFirst[name == jMethod.name]
+            if (foundMethod === null) {
+                createUmlClassMethod(jMethod, uClassifier)
             } else {
-                logger.warn("Can not add ClassMethod to " + uClassifier)
+                addUmlMethodCorrespondence(foundMethod, jMethod)
             }
         }
     }
 }
 
-routine addUmlOperationToClass(uml::Class uClass, uml::Operation uOperation) {
+routine createUmlClassMethod(java::ClassMethod jMethod, uml::Classifier uClassifier) {
     action {
-        update uClass {
-            uClass.ownedOperations += uOperation
+        val uOperation = create uml::Operation and initialize {
+            uOperation.name = jMethod.name;
         }
-    }
-}
-//separate routines because Enum and Class don't have a common super class for "having ownedOperations"
-routine addUmlOperationToEnum(uml::Enumeration uEnum, uml::Operation uOperation) {
-    action {
-        update uEnum {
-            uEnum.ownedOperations += uOperation
+        add correspondence between uOperation and jMethod
+        update uClassifier {
+            (uClassifier as OperationOwner).ownedOperations += uOperation
```

Comment:
I'm not completely sure, why that cast to `OperationOwner` is required, rather than adding the operation to the `operations` attribute of the `Classifier` class. As far as I understand, `OperationOwner` only provides a protocol interface to `Classifier`, especially to have one method for creating an operation without instantiating it on ones own and than adding it to the classifier, but I was not able to understand the difference in accessing the `operations` attribute yet.

Reply:
As discussed: `operations` returns an unmodifiable list, while `ownedOperations` does not.
This behavior is not clear, as the UML documentation does not provide this information.